modules/lang-painless/src/main/java/org/elasticsearch/painless/MethodWriter.java

```diff
@@ -262,6 +262,50 @@ public void writeToStrings() {
         }
     }
 
+    /** Writes a dynamic binary instruction: returnType, lhs, and rhs can be different */
+    public void writeDynamicBinaryInstruction(Location location, Type returnType, Type lhs, Type rhs, Operation operation) {
+        org.objectweb.asm.Type descriptor = org.objectweb.asm.Type.getMethodType(returnType.type, lhs.type, rhs.type);
```

Comment:
Its not easy to refactor without it being trappy. Remember, not all of these are 'binary' instructions. Some are shifts. so each guy needs to set at least 2 things, what type he is *and* what instruction.

To me, we simply have to stop categorizing all of these as the same thing. They are very different operators. It causes all the if-then-else.